File: Python/stage3/S3_PIPE_02.py
# S3_PIPE_02 — compose: right-to-left function composition

def compose(*fns):

    def runner(x):

        for fn in reversed(fns):

            x = fn(x)

        return x

    return runner


# Tests

f = lambda x: x + 1

g = lambda x: x * 2

h = lambda x: x - 3

p = compose(f, g, h)

print(p(10))  # f(g(h(10))) = (10 - 3) * 2 + 1 = 15
 

File: js/stage3/S3_PIPE_02.js
// S3_PIPE_02 — compose (right to left)
const compose =
  (...fns) =>
  (value) =>
    fns.reduceRight((acc, fn) => fn(acc), value);
// Test
const add1 = (x) => x + 1;
const double = (x) => x * 2;
const fn = compose(add1, double);
console.log(fn(3)); // add1(double(3)) = 7
